From 11bcd9477eb485ce7fedee550f80fdd83053fcae Mon Sep 17 00:00:00 2001
From: Yahoo-Mike <ya_mikew@yahoo.com.au>
Date: Mon, 27 Mar 2023 23:11:12 +1100
Subject: [PATCH] Remove a12 kmblob magic.  This device's vendor keymaster
 libraries were written before A12.  In A12 the format of a keymaster blob
 was changed to include an 8 byte magic header.  We must remove the magic so
 that the pre-A12 km can process the kmblob.  This is a re-write to match
 the style of https://gerrit.twrp.me/4623/.

---
 KeyStorage.cpp | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/KeyStorage.cpp b/KeyStorage.cpp
index 6b0584c..531bc67 100644
--- a/KeyStorage.cpp
+++ b/KeyStorage.cpp
@@ -61,6 +61,8 @@ static constexpr size_t GCM_MAC_BYTES = 16;
 static constexpr size_t SECDISCARDABLE_BYTES = 1 << 14;
 constexpr int EXT4_AES_256_XTS_KEY_SIZE = 64;
 
+static const std::string kmBlobMagic("pKMblob\x00", 8);
+
 static const char* kCurrentVersion = "1";
 static const char* kRmPath = "/system/bin/rm";
 static const char* kSecdiscardPath = "/system/bin/secdiscard";
@@ -364,6 +366,10 @@ static KeymasterOperation BeginKeymasterOp(Keymaster& keymaster, const std::stri
         // DeleteUpgradedKey(keymaster, upgraded_blob_file);
     if (!readFileToString(blob_file, &blob)) return KeymasterOperation();
     // }
+    // In A12 keymaster_key_blob format changed.  Vendor libraries before A12 do not
+    // expect the A12 blob magic, so remove any magic to correctly handle key.
+    if (!blob.compare(0, kmBlobMagic.size(), kmBlobMagic))
+        blob.erase(0, kmBlobMagic.size());
     auto opHandle = keymaster.begin(blob, inParams, outParams);
     if (!opHandle) return opHandle;
 
@@ -665,6 +671,10 @@ bool retrieveKey(const std::string& dir, const KeyAuthentication& auth, KeyBuffe
 static bool DeleteKeymasterKey(const std::string& blob_file) {
     std::string blob;
     if (!readFileToString(blob_file, &blob)) return false;
+    // In A12 keymaster_key_blob format changed.  Vendor libraries before A12 do not
+    // expect the A12 blob magic, so remove any magic to correctly handle key.
+    if (!blob.compare(0, kmBlobMagic.size(), kmBlobMagic))
+        blob.erase(0, kmBlobMagic.size());
     Keymaster keymaster;
     if (!keymaster) return false;
     LOG(INFO) << "Deleting key " << blob_file << " from Keymaster";
-- 
2.17.1

